### PR TITLE
Fix scroll leak in scroll handling.

### LINKED ISSUE #1985: Elasticsearch scroll leak
When querying a directIndex with a small number of vertices with unique data at a very high rate the scroll that is created to read the data from ES is not deleted if it fully read in constructor of ElasticSearchScroll.

### PR DESCRIPTION
**Pull request by pkendall64**
If the entire scroll is read in the constructor the scroll was not deleted in ES.

Fixes #1985

Signed-off-by: Paul Kendall <pkendall64@gmail.com>

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there an issue associated with this PR? Is it referenced in the commit message?
- [ ] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [ ] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[doc only]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by FlorianHockmann** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchScroll.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=5068867&bid=16677569)

**Comment by pkendall64** [comment]
When writing unit tests I found another small bug which while not causing problems now, it could potentially if the iterator was not used properly. So I fixed that problem as well and 4 new unit tests for the class.

**Comment by porunov** [comment]
@pkendall64 Thank you! Would you be able to squash your 3 commits into a single commit?

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchScroll.java [hunk 1]
```diff
@@ -28,19 +28,28 @@
...
-    private final BlockingQueue<RawQuery.Result<String>> queue;
+    private final Queue<Result<String>> queue;
     private boolean isFinished;
     private final ElasticSearchClient client;
     private final String scrollId;
     private final int batchSize;
 
     public ElasticSearchScroll(ElasticSearchClient client, ElasticSearchResponse initialResponse, int nbDocByQuery) {
-        queue = new LinkedBlockingQueue<>();
+        queue = new ArrayDeque<>();
```
Comment (porunov) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchScroll.java]: Why did you change this to non blocking queue? What about thread safety?

Comment (pkendall64) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchScroll.java]: I could undo this change as I have fixes the problem in a better way by calling hasNext in the next method to make sure that if a loop is just calling next that it reads all records.
I don't believe that this is used from multiple threads though.

**Comment by FlorianHockmann** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java  2
         

Complexity decreasing per file
==============================
+ janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchScroll.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=5068867&bid=16677569)

File: janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+import io.vavr.collection.List;
+import org.janusgraph.diskstorage.indexing.RawQuery;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.NoSuchElementException;
+
+import static org.easymock.EasyMock.*;
```
Comment (FlorianHockmann) [inline janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.easymock.EasyMock'](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=5068867)

**Comment by FlorianHockmann** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java  2
         

Complexity decreasing per file
==============================
+ janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchScroll.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=5068867&bid=16677569)

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchScroll.java (hunk 1 above)
Comment (pkendall64) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchScroll.java]: Reverted this part of the commit

**Comment by FlorianHockmann** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java  2
         

Complexity decreasing per file
==============================
+ janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchScroll.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=5068867&bid=16677569)

**Comment by pkendall64** [comment]
All squashed into a single commit

**Comment by FlorianHockmann** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java  2
         

Complexity decreasing per file
==============================
+ janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchScroll.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=5068867&bid=16677569)

File: janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
+// Copyright 2017 JanusGraph Authors
```
Comment (farodin91) [inline janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java]: nit: 2020

File: janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        ElasticSearchScroll scroll = new ElasticSearchScroll(client, r1, 5);
+        for (int i=0 ; i<6 ; i++) {
+            scroll.hasNext();
+            scroll.next();
+        }
+
+        verify(client);
+    }
+
+    @Test(expected = NoSuchElementException.class)
```
Comment (farodin91) [inline janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java]: nit: if you know where exception will be thrown use assertThrows.

File: janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.diskstorage.es;
+
+import io.vavr.collection.List;
+import org.janusgraph.diskstorage.indexing.RawQuery;
+import org.junit.Test;
```
Comment (farodin91) [inline janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java]: nit: this should be the junit5 test-method.

File: janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java (hunk 3 above)
Comment (pkendall64) [reply janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java]: fixed

File: janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java (hunk 5 above)
Comment (pkendall64) [reply janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java]: fixed

File: janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java (hunk 4 above)
Comment (pkendall64) [reply janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java]: fixed

**Comment by porunov** [comment]
I have rewritten `ElasticSearchScrollTest` during the merge into the master branch due to change in the response structure. Also, used Mockito only instead of EasyMock. When the tests are finished locally, I will push that merge commit into JanusGraph repository.
Commit: https://github.com/porunov/janusgraph/commit/e10f01b05c4946e9ffa09d46efb114d78cf7ae7d
Tests: https://travis-ci.org/porunov/janusgraph/builds/655878471?utm_source=github_status&utm_medium=notification

File: janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java (hunk 4 above)
Comment (pkendall64) [reply janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java]: done

File: janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java (hunk 5 above)
Comment (pkendall64) [reply janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java]: done

File: janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java (hunk 3 above)
Comment (pkendall64) [reply janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchScrollTest.java]: updated

**Comment by porunov** [comment]
Tests are finished. The merge request is now pushed into the master branch

**Comment by farodin91** [comment]
We have to probably an issue with this PR. https://groups.google.com/forum/#!topic/janusgraph-users/Bv_tYFI38kc

**Comment by porunov** [comment]
Yes. I will fix it soon. If we don't use scrolling because our limit is less than batch, then scroll is null. We should delete scroll only if it isn't null

**Pull request merged**